test/value.cpp

```diff
@@ -2087,6 +2087,38 @@ class value_test
             {"a",1}, {"b",2}, {"c",3} }) != value(nullptr));
     }
 
+    void
+    testHash()
+    {
+        auto hnull = std::hash<value>{}(value(nullptr));
+        auto htrue = std::hash<value>{}(value(true));
+        auto hfalse = std::hash<value>{}(value(false));
+        auto hi0 = std::hash<value>{}(value(2021));
+        auto hu0 = std::hash<value>{}(value(2021U));
+        auto hd0 = std::hash<value>{}(value(2021.0));
+
+        auto hobj0 = std::hash<value>{}(value({{"a",1}, {"b",2}, {"c",3}}));
+        auto hobj1 = std::hash<value>{}(value({{"b",2}, {"c",3}, {"a",1}}));
+        auto hobj2 = std::hash<value>{}(value({{"b",2}, {"c",3}}));
+
+        auto harr0 = std::hash<value>{}(value{"a", "b", 17});
+        auto harr1 = std::hash<value>{}(value{"a", "b", 17U});
+        auto harr2 = std::hash<value>{}(value{17, "a", "b"});
+
+        BOOST_TEST(hnull != htrue);
+        BOOST_TEST(hnull != hfalse);
+        BOOST_TEST(htrue != hfalse);
+        BOOST_TEST(hi0 == hu0);
+        BOOST_TEST(hi0 != hd0);
+
+        BOOST_TEST(hobj0 == hobj1);
+        BOOST_TEST(hobj0 != hobj2);
+
+        BOOST_TEST(harr0 == harr1);
+        BOOST_TEST(harr1 != harr2);
```

Comment:
Nice test coverage, but lets be honest here - there would have been a small but non zero chance that these hashes collided :)

Reply:
2^-32 isn't bad.

I'd have written this differently - with a helper function that when v1 == v2 checks that the hashes match, and when v1 != v2 checks that they differ, so as to test the correspondence between op== and the hash. Or with two helper functions, one checking the equal case and one checking the unequal case.

Or I suppose one can just check equality inline.

Things like 0 != false, {} != 0, [] != 0, {} != [], [] != [[]], etc should be checked, too. The above doesn't check across kinds (except for null).

Reply:
Will update tests accordingly.